src/hooks/usePipelineRuns.ts

```diff
@@ -0,0 +1,271 @@
+import * as React from 'react';
+import {
+  K8sGroupVersionKind,
+  K8sResourceCommon,
+  Selector,
+  useK8sWatchResource,
+} from '@openshift/dynamic-plugin-sdk-utils';
+import { differenceBy, uniqBy } from 'lodash-es';
+import { PipelineRunLabel, PipelineRunType } from '../consts/pipelinerun';
+import { PipelineRunGroupVersionKind } from '../models';
+import { useDeepCompareMemoize } from '../shared';
+import { PipelineRunKind, TaskRunGroupVersionKind, TaskRunKind } from '../types';
+import { EQ } from '../utils/tekton-results';
+import { GetNextPage, useTRPipelineRuns, useTRTaskRuns } from './useTektonResults';
+
+const useRuns = <Kind extends K8sResourceCommon>(
+  groupVersionKind: K8sGroupVersionKind,
+  namespace: string,
+  options?: {
+    selector?: Selector;
+    limit?: number;
+    name?: string;
+  },
+): [Kind[], boolean, unknown, GetNextPage] => {
+  const etcdRunsRef = React.useRef<Kind[]>([]);
+  const optionsMemo = useDeepCompareMemoize(options);
+  const limit = optionsMemo?.limit;
+  const isList = !optionsMemo?.name;
+  // do not include the limit when querying etcd because result order is not sorted
+  const watchOptions = React.useMemo(() => {
+    // reset cached runs as the options have changed
+    etcdRunsRef.current = [];
+    return namespace
+      ? {
+          groupVersionKind,
+          namespace,
+          isList,
+          selector: optionsMemo?.selector,
+          name: optionsMemo?.name,
+        }
+      : null;
+  }, [groupVersionKind, namespace, optionsMemo, isList]);
+
+  const [resources, loaded, error] = useK8sWatchResource(watchOptions);
+  // if a pipeline run was removed from etcd, we want to still include it in the return value without re-querying tekton-results
+  const etcdRuns = React.useMemo(
+    () => (isList ? resources : loaded ? [resources] : []) as Kind[],
```

Comment:
```suggestion
    () => (isList ? resources : loaded && !error ? [resources] : []) as Kind[],
```

Reply:
it's very odd that `useK8sWatchResource` can return `[{}, true, error]`. Why an empty object when we have an error.
I've added a test case for this and updated to the suggestion. All other test cases still pass.